### ISSUE TITLE
Operations usage in SEES and USES seems to be incorrect

### ISSUE THREAD

**Issue by SeeBasTStick** [hhu-stups/prob-issues#84]
While experimenting with the different forms of inclusion unexpected behavior occurred. 

Consider the following scenario:
```
MACHINE MachineA

VARIABLES x
INVARIANT x : 1..10
INITIALISATION x := 1
OPERATIONS 
  Inc = SELECT x < 10 THEN x := x + 1 END;
  out <-- Get = out := x
END
```
```
MACHINE MachineB

USES MachineA
VARIABLES y
INVARIANT y : NAT
INITIALISATION y := 10

OPERATIONS 
 Foo = Inc || y := 10 ;
 Bar = y <-- Get   
END
```
This causes ```query-operation expected ```. Therefore only read operations are permitted. According to AtelierB and the handbook, read operations are not permitted in ```USES``` but they are allowed in ```SEES```. 

Using ProB2 Animator to animate the Machine2 (removing the first incorrect operation), in both cases, ```USES``` and ```SEES```,  the animator exposes the operations of the underlying machine as if they were promoted which is not according to the handbook. However, I am not sure if this is intended. What could be concerning tho is if this exposer means that the paths are checked in a model checking process. As they are incorrect model checking could be also compromised by finding false positives.

TL;DR
* Change the message from ```query-operation expected``` to a one that is similar to the one In AtelierB: ``` Read-Only Operation Expected```
* ```SEES``` can use read-only operation, ```USES``` not
* In both cases the operations shouldn't be exposed in the animator as if they were promoted.

**Comment by leuschel** [comment]
I have improved the query-operation error message; now it says:
```
! An error occurred (source: type_error) !
! Query-operation expected (operation Inc writes, e.g., variable x)
```

**Comment by leuschel** [comment]
actually it now is
```
! Read-Only operation expected (operation Inc writes, e.g., variable x)
```

**Comment by leuschel** [comment]
The fact that the operations of used machines are virtually promoted is on purpose. A USEd machine is controlled by somebody else. So MachineA is *not* under the control of MachineB and between operation calls to MachineB an arbitrary number of operations of MachineA can be executed (by the real master of MachineA). Hence, the virtual promotion of Inc is actually important to catch errors in MachineB.

**Comment by leuschel** [comment]
In Schneider's B Book it says "The using machine has read access to M1 in exactly the same way as provided by sees. ... can also refer to the state of M1 within its invariant." So if one call a query-operation via SEES, one can also call it via USES. But it is true that the Handbook seems to forbid calling query operations of a used machine. I am not sure why, as all variables, constants, parameters are visible to operations of the using machine; hence I think one could write a copy of the query-operation.

**Comment by SeeBasTStick** [comment]
> The fact that the operations of used machines are virtually promoted is on purpose. A USEd machine is controlled by somebody else. So MachineA is _not_ under the control of MachineB and between operation calls to MachineB an arbitrary number of operations of MachineA can be executed (by the real master of MachineA). Hence, the virtual promotion of Inc is actually important to catch errors in MachineB.

Okay to recapitulated: The reason for the ability to use these virtual promoted operations is to catch the fact that the USEing machine is not the real master and only USEs the machine, while the real master of the USEd machine can do whatever he/she likes. 

Currently I have a hard time to think about an example when this virtual promotion is actual useful.

What I would suggest as an improvement would be to mark these virtual promotions with some extra color in ProB2Ui. It does not make a difference for MC but it indeed does for animation purposes.

**Comment by leuschel** [comment]
Quite often you need the virtual promotion in the animation to change the state of the used machine to a desired value. This feature avoids the need to write a dummy master machine which includes both A and B.

**Comment by leuschel** [comment]
Commit 400ff21724112ffc06340c00c99693a344176270 now disallows calling operations of a USEd machine.

**Comment by SeeBasTStick** [comment]
> Quite often you need the virtual promotion in the animation to change the state of the used machine to a desired value. This feature avoids the need to write a dummy master machine which includes both A and B.

This makes totally sense, I get it know - thank you. I think that is a good reason to distinguish virtual and real promoted operations visually. However, it might not be that important.

**Comment by leuschel** [comment]
Figure 11.7 on page 166 in the B-Book by Schneider clearly shows that calling of query operations is allowed. I guess the BToolkit did allow this, and Atelier-B not. As ProB tries to conform to Atelier-B, I have changed the setting. One could introduce a preference if needed.

**Comment by leuschel** [comment]
There is now a preference ALLOW_OPERATION_CALLS_FOR_USES which can be used to obtain the previous behaviour as described in Schneider's book and I guess BToolkit
